pkg/reconciler/pipelinerun/pipelinerun_test.go

```diff
@@ -3759,3 +3305,112 @@ func getTaskRunStatus(t string, status corev1.ConditionStatus) *v1beta1.Pipeline
 		},
 	}
 }
+
+func buildTestData(
+	prs []*v1beta1.PipelineRun,
+	ps []*v1beta1.Pipeline,
+	ts []*v1beta1.Task,
+	clusterTasks []*v1beta1.ClusterTask,
+	trs []*v1beta1.TaskRun,
+	rs []*v1alpha1.PipelineResource,
+	conditions []*v1alpha1.Condition,
+) test.Data {
+	return test.Data{
+		PipelineRuns:      prs,
+		Pipelines:         ps,
+		TaskRuns:          trs,
+		Tasks:             ts,
+		ClusterTasks:      clusterTasks,
+		PipelineResources: rs,
+		Conditions:        conditions,
+	}
+}
+
+func getReconciledRun(
+	t *testing.T,
+	prs []*v1beta1.PipelineRun,
+	ps []*v1beta1.Pipeline,
+	ts []*v1beta1.Task,
+	clusterTasks []*v1beta1.ClusterTask,
+	trs []*v1beta1.TaskRun,
+	rs []*v1alpha1.PipelineResource,
+	conditions []*v1alpha1.Condition,
+	namespace, prName string,
+	wantEvents []string,
+) (*v1beta1.PipelineRun, test.Clients) {
+
+	d := buildTestData(prs, ps, ts, clusterTasks, trs, rs, conditions)
+
+	testAssets, cancel := getPipelineRunController(t, d)
+	defer cancel()
+	c := testAssets.Controller
+	clients := testAssets.Clients
+	recorder := testAssets.Recorder
+
+	if reconcileError := c.Reconciler.Reconcile(context.Background(), namespace+"/"+prName); reconcileError != nil {
+		t.Fatalf("Error reconciling: %s", reconcileError)
```

Comment:
have you considered using t.Helper? (https://blog.golang.org/go1.9)

It might be worth comparing what it looks like when one of the tests using this helper fails using it vs not using it (it might make more sense if you were to break this function up into a few)

Reply:
Yup, very useful, instead of pointing back at the common utility (`reconcileRun`), it points back where the failure is originated 👍